File: Data Structures & Algorithms/house-robber/submission-1.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n);

        dp[0] = nums[0];
        dp[1] = nums[1];
        if (nums.size() == 1){
            return dp[0];
        }
        else if (nums.size() == 2) {
            return max(dp[0], dp[1]);
        }

        for (int i = 2; i < n; i++) {
            int maxVal = 0;
            for (int j = 0; j < i - 1; j++) {
                maxVal = max(maxVal, dp[j] + nums[i]);
            }
            dp[i] = max(maxVal, dp[i - 1]);
        }

        return dp[n - 1];
    }
};

File: Data Structures & Algorithms/house-robber-ii/submission-1.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 1:
            return nums[0]

        rob1, rob2, rob3, rob4 = 0, 0, 0, 0

        for i in range(n - 1):
            rob1, rob2 = rob2, max(rob2, rob1 + nums[i])
            rob3, rob4 = rob4, max(rob4, rob3 + nums[i + 1])
        
        return max(rob2, rob4)